### PR TITLE
Support composing getter/accessor types

### PR DESCRIPTION
**Pull request by blitz-1306**
## Premise
This PR is a part of logic to address https://github.com/ConsenSys/scribble/issues/75.

## Changes
- [x] Tweaks for `assert()` to support placeholders. This would allow to delay heavy logic (like `ASTNode.print()` or `TypeNode.pp()`) until (and if) the assertion is failed and there is a real necessity to throw an error (with composed message).
- [x] Introduced `typeNameToSpecializedTypeNode()` shorthand to regularly used `specializeType(typeNameToTypeNode(node), loc)`.
- [x] Introduced `enumToIntType()`  to convert enum definition types to `int` value types.
- [x] Introduced `getterArgsAndReturn()` and `getterTypeForVar()` to implement requested functionality.
- [x] Tweaked `EnumDefinition.toUintTypeString()` to use `enumToIntType()`.
- [x] Other performance and readability tweaks in type-related utils.
- [x] Added related tests.

Regards.

### CODE REVIEW

File: src/types/utils.ts [hunk 1]
```diff
@@ -257,3 +245,152 @@ export function typeNameToTypeNode(astT: TypeName): TypeNode {
...
+ * 2. Enum definitions turned to uint of minimal fitting size.
+ * 3. Struct definitions turned to tuples.
+ *   3.1. Mappings are completely ommited.
+ *   3.2. Arrays of root structs are ommited, while preserved for nested structs.
+ * 4. Data locations are skipped.
+ */
+function toValueType(type: TypeNode, depth = 0): TypeNode {
+    if (type instanceof PointerType) {
+        return toValueType(type.to, depth);
+    }
```
Comment (blitz-1306) [inline src/types/utils.ts:301]: @cd1m0 I wonder if this is valid. Maybe we should skip `specializeType()` in downstream instead.

File: test/samples/solidity/getters.sol [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    struct S2 {
+        uint[] a;
+        address b;
+    }
+
+    struct S {
+        uint[] a;
+        int8 b;
+        string x;
+        S1 s1;
```
Comment (blitz-1306) [inline test/samples/solidity/getters.sol:25]: @cd1m0 This is very interesting edge-case. Nested tuples preserve arrays, while root structs do not.

File: src/types/utils.ts (hunk 1 above)
Comment (cd1m0) [reply src/types/utils.ts:301]: I would prefer to generalize types here, and keep types "specialized" everywhere else. Both maru-ir and scribble depend on `variableDeclarationToTypeNode` to return a specialized type, and they specialize the return of `typeNameToTypeNode` whenever they use it. It seems that we only generalize types when we use them for something web3 related

File: test/samples/solidity/getters.sol (hunk 2 above)
Comment (cd1m0) [reply test/samples/solidity/getters.sol:25]: Yep. I also noticed this one

File: src/types/utils.ts [hunk 3]
```diff
@@ -257,3 +245,152 @@ export function typeNameToTypeNode(astT: TypeName): TypeNode {
...
+    );
+
+    while (true) {
+        if (type instanceof ArrayTypeName) {
+            argTypes.push(new IntType(256, false));
+
+            type = type.vBaseType;
+        } else if (type instanceof Mapping) {
+            argTypes.push(
+                toValueType(typeNameToSpecializedTypeNode(type.vKeyType, DataLocation.Memory))
```
Comment (cd1m0) [inline src/types/utils.ts:369]: Why are we calling `toValueType` here? I would think that if we are computing the type for a getter for var `mapping(string => uint)` the argument would be `string memory` not `string`?

**Review by cd1m0 — CHANGES_REQUESTED** [review]
One question. Otherwise lgtm

File: src/types/utils.ts (hunk 3 above)
Comment (blitz-1306) [reply src/types/utils.ts:369]: Due to `Enum`s possible as mapping keys.

Comment (blitz-1306) [reply src/types/utils.ts:384]: Regarding `string memory`, see https://github.com/ConsenSys/solc-typed-ast/commit/83cda55cac6638e4dd593bf56869eb33b22584d4. Allowed non-definition types to be pointers. Also pointers are reconstructed to point to `memory` data location.

File: src/types/utils.ts [hunk 4]
```diff
@@ -269,3 +262,179 @@ export function typeNameToTypeNode(astT: TypeName): TypeNode {
...
+
+    return new IntType(size, false);
+}
+
+/**
+ * Converts `type` to value type. Following rules are applied:
+ * 1. Contract definitions turned to address.
+ * 2. Enum definitions turned to uint of minimal fitting size.
+ */
+function toValueType(type: TypeNode, encoderVersion: ABIEncoderVersion): TypeNode {
```
Comment (cd1m0) [inline src/types/utils.ts:311]: The name of the function is a bit general. Maybe something like `toAccessorType`? Not sure. Also not a blocker, might merge as is.

File: test/samples/solidity/getters_07.sol [hunk 5]
```diff
@@ -0,0 +1,29 @@
+pragma solidity ^0.7.0;
```
Comment (cd1m0) [inline test/samples/solidity/getters_07.sol:1]: As a regression test we might want to also add an ABIEncoderV1 test. Not a blocker

**Review by cd1m0 — APPROVED** [review]
LGTM. Minor nits that can be fixed in a followup PR

**Review by cd1m0 — CHANGES_REQUESTED** [review]
So after playing around I found another issue. So there is a difference between what the signature of accessors are at the _Solidity_ level, and at the level of the contract's _ABI_. Specifically contracts returned from public vars are not treated as addresses at the solidity level, just at the API level. The same is true for enums - at the solidity levels they are treated as enums, not uints. This samples shows the failures:

```solidity
pragma solidity ^0.8.0;
contract T {}

contract Foo {
    enum E {A,B,C}
    
    T public t;
    E public e;
    
    function main() public {
        address a = this.t();
        uint8 b = this.e();
    }
}
```

Both assignments in main will show type errors. One way to fix this is to:

1. Change `getterArgsAndReturn` to not use `toValueType`, and instead just convert non-memory pointer types to memory pointers
2. Add another `getterExternalArgsAndReturn` which calls `getterArgsAndReturn`, and calls `toValueType` on it.

Also we have code duplication with `VariableDeclaration.canonicalSignatureType` and `VariableDeclaration.getterCanonicalSignature`. First of all, not sure why we have two separate functions? I think they should do the same. And second, I think they should just call `getterExternalArgsAndReturn`

**Comment by blitz-1306** [comment]
@cd1m0 I guess we somehwere went too far with public getters on ABI | internal level. Instead of complicating the code base I think we can remove `toValueType()` and dependent parts, just leave the `getterTypeForVar()` and `getterArgsAndReturn()` to build internal getter function type. Then adding a logic to convert it to ABI representation should be easier and may be left to consumers (or maybe done as a followup). I also changed approach for the tests to replicate actual function signatures (it looks more transparent). See https://github.com/ConsenSys/solc-typed-ast/pull/64/commits/bf45d33711e23ef99727b13d3141fb6bcd8d7559.

File: test/samples/solidity/getters_07.sol (hunk 5 above)
Comment (cd1m0) [reply test/samples/solidity/getters_07.sol:1]: Its added in my followup PR so no need to do this here.

**Review by cd1m0 — APPROVED** [review]
LGTM!

File: src/types/utils.ts (hunk 4 above)
Comment (cd1m0) [reply src/types/utils.ts]: Irrelevant since bf45d33
